src/XMakeBuildEngine/Construction/ProjectRootElement.cs

```diff
@@ -1856,6 +1870,84 @@ public ProjectRootElement DeepClone()
+        /// 
+        /// See <see cref="ProjectRootElement.ReloadFrom(XmlReader, bool, bool?)"/>
+        /// </summary>
+        public void ReloadFrom(string path, bool throwIfUnsavedChanges = true, bool? preserveFormatting = null)
+        {
+            ErrorUtilities.VerifyThrowInvalidOperation(File.Exists(path), "FileToReloadFromDoesNotExist", path);
+
+            Func<bool, XmlDocumentWithLocation> documentProducer = shouldPreserveFormatting => LoadDocument(path, shouldPreserveFormatting);
+            ReloadFrom(documentProducer, throwIfUnsavedChanges, preserveFormatting);
+        }
+
+        /// <summary>
+        /// Reload the existing project root element from the given <paramref name="reader"/>
+        /// A reload operation completely replaces the state of this <see cref="ProjectRootElement"/> object. This operation marks the 
+        /// object as dirty (see <see cref="ProjectRootElement.MarkDirty"/> for side effects). 
+        /// 
+        /// If the new state has invalid XML or MSBuild syntax, then this method throws an <see cref="InvalidProjectFileException"/>.
+        /// When this happens, the state of this object does not change.
+        /// 
+        /// </summary>
+        /// <param name="reader">Reader to read from</param>
+        /// <param name="throwIfUnsavedChanges">
+        ///   If set to false, the reload operation will discard any unsaved changes.
+        ///   Otherwise, an <see cref="InvalidOperationException"/> is thrown when unsaved changes are present.
+        /// </param>
+        /// <param name="preserveFormatting">
+        ///   Whether the reload should preserve formatting or not. A null value causes the reload to reuse the existing <see cref="PreserveFormatting"/> value.
+        /// </param>
+        public void ReloadFrom(XmlReader reader, bool throwIfUnsavedChanges = true, bool? preserveFormatting = null)
+        {
+            Func<bool, XmlDocumentWithLocation> documentProducer = shouldPreserveFormatting => LoadDocument(reader, shouldPreserveFormatting);
+            ReloadFrom(documentProducer, throwIfUnsavedChanges, preserveFormatting);
+        }
+
+        private void ReloadFrom(Func<bool, XmlDocumentWithLocation> documentProducer, bool throwIfUnsavedChanges, bool? preserveFormatting)
+        {
+            ThrowIfUnsavedChanges(throwIfUnsavedChanges);
+
+            XmlDocumentWithLocation document = documentProducer(preserveFormatting ?? PreserveFormatting);
+
+            // Reload should only mutate the state if there are no parse errors.
+            ThrowIfDocumentHasParsingErrors(document);
+
+            // Do not clear the string cache.
+            // Based on the assumption that Projects are reloaded repeatedly from their file with small increments,
+            // and thus most strings would get reused
+            //this.XmlDocument.ClearAnyCachedStrings();
+
+            this.RemoveAllChildren();
+
+            ProjectParser.Parse(document, this);
```

Comment:
The parser mutates the state of the root element (sets both XmlNode and ProjectElement elements) while it's parsing, so if it encounters an error you get a partially mutated root element. So we'll have to manually copy over the state from the test root element. Left it as a todo since it's as fast as the previous usage which also parsed twice, and it's worth doing after we flesh out the other aspects of Reload.